File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int helper(vector<int>& nums) {
        int prev=nums[0];
        int prev2=0;
        for(int i=1;i<nums.size();i++){
            int take=nums[i];
            if(i>1){
                take+=prev2;
            }
            int nottake=0+prev;
            int c=max(take,nottake);

            prev2=prev;
            prev=c;
        }
        return prev;
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int>v1,v2;
        if(n==1){
            return nums[0];
        }
        for(int i=0;i<n;i++){
            if(i!=0){
                v1.push_back(nums[i]);
            }
            if(i!=n-1){
                v2.push_back(nums[i]);
            }
        }
        int mx1=helper(v1);
        int mx2=helper(v2);
        return max(mx1,mx2);
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {


    public int rob(int[] nums) {
        int n=nums.length;
        if(n==1){
            return nums[0];
        }
        if(n==2){
            return Math.max(nums[0],nums[1]);
        }

        int[] dp=new int[n];
        int[] dp1=new int[n];

        dp[0]=nums[0];
        dp[1]=Math.max(nums[0],nums[1]);

        dp1[1]=nums[1];
        dp1[2]=Math.max(nums[2],nums[1]);

        for(int i=2;i<=n-2;i++){
            dp[i]=Math.max(nums[i]+dp[i-2],dp[i-1]);
        }

        for(int i=3;i<=n-1;i++){
            dp1[i]=Math.max(nums[i]+dp1[i-2],dp1[i-1]);
        }

        

        return Math.max(dp[n-2],dp1[n-1]);
    }
}